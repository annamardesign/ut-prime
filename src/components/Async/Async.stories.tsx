import React from 'react';
import type { Meta } from '@storybook/react';

// @ts-ignore: md file and not a module
import page from './README.mdx';
import Async from './index';

const meta: Meta = {
    title: 'Async',
    component: Async,
    parameters: {docs: {page}}
};
export default meta;

const Loaded = () => <div className='p-component'>async component loaded</div>;
const Delay = () => new Promise<React.FC>((resolve, reject) => {
    setTimeout(() => resolve(Loaded), 1000);
});

export const Basic: React.FC<{}> = () => <Async component={Delay} />;
