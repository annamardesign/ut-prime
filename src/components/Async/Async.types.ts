import React from 'react';

export type asyncComponentParams = {}
export type asyncComponent = (params: asyncComponentParams) => Promise<React.FC>

export interface Props {
    params?: asyncComponentParams,
    search?: string,
    component: asyncComponent
};

export type StyledType = React.FC<Props>
