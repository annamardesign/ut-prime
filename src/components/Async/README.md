# Async

Renders a component returned by an async function.
Alternative to React.Suspense and React.lazy.

## How to use

```jsx
import Async from 'ut-prime/core/Async';

<Async component={asyncFn} />;
```

## Props

- **className** - class applied to root `Async` html element.

| propName  | propType | defaultValue | isRequired |
| --------- | -------- | ------------ | ---------- |
| component | function |              | yes         |
