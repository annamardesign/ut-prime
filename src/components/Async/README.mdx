# Async
import { Canvas, Story, ArgsTable } from "@storybook/addon-docs";

Renders a component returned by an async function.
Alternative to React.Suspense and React.lazy.

## How to use

```jsx
import Async from 'ut-prime/core/Async';

<Async component={asyncFn} />;
```

<Canvas>
    <Story id="async--basic" />
</Canvas>

## Properties

<ArgsTable story="."/>
