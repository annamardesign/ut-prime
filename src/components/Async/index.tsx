import React from 'react';

import { StyledType, asyncComponent, asyncComponentParams } from './Async.types';

// adapted from https://usehooks.com/useAsync/
const useAsync = (asyncFunction: asyncComponent, params: asyncComponentParams) => {
    const [status, setStatus] = React.useState<'idle' | 'pending' | 'success' | 'error'>('idle');
    const [value, setValue] = React.useState<{component: React.FC}>(null);
    const [error, setError] = React.useState<Error>(null);

    // The execute function wraps asyncFunction and
    // handles setting state for pending, value, and error.
    // useCallback ensures the below useEffect is not called
    // on every render, but only if asyncFunction changes.
    const execute = React.useCallback(async() => {
        setStatus('pending');
        setValue(null);
        setError(null);
        try {
            setValue({component: await asyncFunction(params)});
            setStatus('success');
        } catch (error) {
            setError(error);
            setStatus('error');
        }
    }, [asyncFunction, params]);

    React.useEffect(() => {
        execute();
    }, [execute]);

    return { execute, status, Component: value && value.component, error };
};

const Async: StyledType = ({component, params, children}) => {
    const { status, Component, error } = useAsync(component, params);
    switch (status) {
        case 'error': return <div>{JSON.stringify(error)}</div>;
        case 'success': return <Component {...params}>{children}</Component>;
        default: return <div>Loading...</div>;
    }
};

export default Async;
