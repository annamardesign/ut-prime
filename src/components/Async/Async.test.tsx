import React from 'react';
import { act } from 'react-dom/test-utils';

import { render } from '../test';
import Async from './index';

describe('<Async />', () => {
    it('render equals snapshot', async() => {
        let res: (component: React.FC) => void;
        const success = () => new Promise<React.FC>(resolve => { res = resolve; });
        const { getByTestId } = render(<Async component={success} />);
        expect(getByTestId('ut-front-test')).toMatchSnapshot('loading');
        await act(async() => {
            res(() => <div>resolved</div>);
        });
        expect(getByTestId('ut-front-test')).toMatchSnapshot('loaded');
    });
});
