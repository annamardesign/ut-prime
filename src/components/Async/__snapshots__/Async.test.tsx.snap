// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Async /> render equals snapshot: loaded 1`] = `
<div
  data-testid="ut-front-test"
>
  <div>
    resolved
  </div>
</div>
`;

exports[`<Async /> render equals snapshot: loading 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="p-component"
  >
    Loading...
  </div>
</div>
`;
